src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -70,22 +74,23 @@ internal class ProjectBuildManager
             }
         }
 
-        public ProjectBuildManager(ImmutableDictionary<string, string> additionalGlobalProperties)
+        public ProjectBuildManager(ImmutableDictionary<string, string> additionalGlobalProperties, ILogger? msbuildLogger = null)
         {
             _additionalGlobalProperties = additionalGlobalProperties ?? ImmutableDictionary<string, string>.Empty;
+            _msbuildLogger = msbuildLogger;
         }
 
         private ImmutableDictionary<string, string> AllGlobalProperties
             => s_defaultGlobalProperties.AddRange(_additionalGlobalProperties);
 
-        private static async Task<(MSB.Evaluation.Project project, DiagnosticLog log)> LoadProjectAsync(
-            string path, MSB.Evaluation.ProjectCollection projectCollection, CancellationToken cancellationToken)
+        private static async Task<(MSB.Evaluation.Project? project, DiagnosticLog log)> LoadProjectAsync(
+            string path, MSB.Evaluation.ProjectCollection? projectCollection, CancellationToken cancellationToken)
```

Comment:
When could projectCollection ever be null?

Reply:
its nulled out and re-allocated at the start of each Start/End build. so it can easily be null if things are ever called slightly out of order. How does roslyn typically handle complex lifetimes like this that the compiler can't track? I was inclined to say that if your object has a complex lifetime beyond some kind of `Initialize` method then it should be considered nullable. Happy to refactor this so the lifetime isn't so complex though.

---
In reply to: [401260471](https://github.com/dotnet/roslyn/pull/42319#discussion_r401260471) [](ancestors = 401260471)